File: awc-2021/p6.py
# Logic was built here - TLEs on Codedrills


def find(dsu, i):
    if dsu[i] != i:
        dsu[i] = find(dsu, dsu[i])
    return dsu[i]

def join(dsu, sz, i, j):
    a = find(dsu, i)
    b = find(dsu, j)
    if a != b:
        if sz[a] < sz[b]:
            a, b = b, a
        sz[a] += sz[b]
        dsu[b] = dsu[a]
        return 1
    return 0

for _ in range(int(input())):
    n, m = map(int, input().split())
    c = []
    valid = 1
    for i in range(n):
        c.append(list(map(int, input().split())))
        if c[-1][0] != 1:
            valid = 0
    if not valid:
        print(-1)
        continue
    
    # graph = {i:[] for i in range(n)}
    edges = []
    dsus = [[i for i in range(n-j)] for j in range(n)]
    szs = [[1 for i in range(n-j)] for j in range(n)]
    ccs = [1 for _ in range(n)]
    edges = set()
    potential_edges = set()
    non_edges = set()
    for k in range(1, n):
        for i in range(n-k):
            ccs[i] += 1 # assume not connected
            for j in range(k-1): # check all the things other than i that i+k is connected to
                if find(dsus[i+1], j) == find(dsus[i+1], k-1):
                    
                    ccs[i] -= join(dsus[i], szs[i], j+1, k)
                
            if c[i][k] != ccs[i]:     
                ccs[i] -= join(dsus[i], szs[i], 0, k)
                edges.add((i+1, i+k+1))
            elif find(dsus[i], 0) != find(dsus[i], k):
                non_edges.add((i+1, i+k+1))
            if c[i][k] != ccs[i]:
                valid = 0
    
    for i in range(n):
        for j in range(i):
            if len(edges) < m and (j+1, i+1) not in non_edges:
                edges.add((j+1, i+1))
    
    if not valid or len(edges) != m:
        print(-1)
    else:
        edges = list(edges)
        edges.sort()
        for edge in edges:
            print(*edge)

File: awc-2021/p6.cpp
#include <iostream>
#include <vector>
#include <set>
#include <algorithm>

using namespace std;

int find(vector<int>& dsu, int i) {
    if (dsu[i] != i) {
        dsu[i] = find(dsu, dsu[i]);
    }
    return dsu[i];
}

bool join(vector<int>& dsu, vector<int>& sz, int i, int j) {
    int a = find(dsu, i);
    int b = find(dsu, j);
    if (a != b) {
        if (sz[a] < sz[b]) swap(a, b);
        sz[a] += sz[b];
        dsu[b] = a;
        return true;
    }
    return false;
}

int main() {
    int t;
    cin >> t;
    while (t--) {
        int n, m;
        cin >> n >> m;
        vector<vector<int>> c(n, vector<int>(n));
        bool valid = true;
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n - i; j++) {
                cin >> c[i][j];
            }
            if (c[i][0] != 1) valid = false;
        }
        
        if (!valid) {
            cout << "-1\n";
            continue;
        }
        
        vector<vector<int>> dsus(n);
        vector<vector<int>> szs(n);
        vector<int> ccs(n, 1);
        set<pair<int, int>> edges, non_edges;
        
        for (int j = 0; j < n; j++) {
            dsus[j].resize(n - j);
            szs[j].resize(n - j, 1);
            for (int i = 0; i < n - j; i++) {
                dsus[j][i] = i;
            }
        }
        
        for (int k = 1; k < n; k++) {
            for (int i = 0; i < n - k; i++) {
                ccs[i]++;
                for (int j = 0; j < k - 1; j++) {
                    if (find(dsus[i + 1], j) == find(dsus[i + 1], k - 1)) {
                        ccs[i] -= join(dsus[i], szs[i], j + 1, k);
                    }
                }
                
                if (c[i][k] != ccs[i]) {
                    ccs[i] -= join(dsus[i], szs[i], 0, k);
                    edges.insert({i + 1, i + k + 1});
                } else if (find(dsus[i], 0) != find(dsus[i], k)) {
                    non_edges.insert({i + 1, i + k + 1});
                }
                
                if (c[i][k] != ccs[i]) valid = false;
            }
        }
        
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < i; j++) {
                if (edges.size() < (size_t)m && non_edges.find({j + 1, i + 1}) == non_edges.end()) {
                    edges.insert({j + 1, i + 1});
                }
            }
        }
        
        if (!valid || edges.size() != (size_t)m) {
            cout << "-1\n";
        } else {
            vector<pair<int, int>> edge_list(edges.begin(), edges.end());
            sort(edge_list.begin(), edge_list.end());
            for (auto& edge : edge_list) {
                cout << edge.first << " " << edge.second << "\n";
            }
        }
    }
    return 0;
}